### PR TITLE
Add new FlawAcknowledgment API

### PR DESCRIPTION
**Pull request by jsvob**
Follow-up to https://github.com/RedHatProductSecurity/osidb/pull/250.

Closes OSIDB-1002 ("Implement new API for FlawAcknowledgment")

### CODE REVIEW

File: openapi.yml
Comment (osoukup) [inline openapi.yml:6154]: I may be wrong so take this just as something to check but is it correct to have `flaw` and `uuid` as required attributes when they are parts of the path? Just seems to be duplicite but I might miss something.

Comment (osoukup) [reply openapi.yml:6154]: Probably good to check with the bindings - FYI @JakubFrejlach

**Review by osoukup — APPROVED** [review]
LGTM

File: openapi.yml
Comment (JakubFrejlach) [reply openapi.yml:6154]: Agree with @osoukup, both PUT and POST should not require flaw ID inside the request body since it is already provided in the path.

File: osidb/api_views.py
Comment (JakubFrejlach) [inline osidb/api_views.py:371]: This queryset won't take into account the `flaw_id` path parameter and will include all acknowledgements. Similar handling as for Flaw Comments or Flaw References should be done.

File: osidb/api_views.py
Comment (JakubFrejlach) [inline osidb/api_views.py:370]: This ViewSet does not have the destroy method overridden and thus the delete will be perform only on the OSIDB side and not on the BZ side. Similar handling as for Flaw Affects or Flaw References should be done.

Also I wonder whether it does make sense to create some common mixin which would already provide the overridden delete.

File: osidb/tests/test_endpoints.py
Comment (JakubFrejlach) [inline osidb/tests/test_endpoints.py:1631]: Should be already included in the path.

File: osidb/api_views.py
Comment (JakubFrejlach) [inline osidb/api_views.py:370]: Destroy method should also return code 200 instead of 204 because we inject some default info about OSIDB like:

```
  "dt": "2023-07-03T07:22:10.661Z",
  "env": "string",
  "revision": "string",
  "version": "string"
  ```

and 204 response by standard won't include any response body.

**Review by JakubFrejlach — DISMISSED** [review]
Couple of things which needs to be addressed.

File: osidb/serializer.py
Comment (jobselko) [inline osidb/serializer.py:627]: `FlawAcknowledgmentSerializer` should inherit also from `IncludeMetaAttrMixin`.

File: osidb/serializer.py
Comment (jobselko) [inline osidb/serializer.py:646]: `flaw` should be also excluded

File: osidb/filters.py
Comment (jobselko) [inline osidb/filters.py:205]: I am unsure whether we need all these filter fields (e.g. `"acknowledgments__created_dt": ["exact"]` does not make sense to me). @JakubFrejlach Could you please look at it?

**Review by jobselko — CHANGES_REQUESTED** [review]

**Comment by jobselko** [comment]
This PR should have an entry in CHANGELOG because it contains changes visible to customers.

File: osidb/filters.py
Comment (JakubFrejlach) [reply osidb/filters.py:205]: Well, it is unlikely that someone will need to filter out exact datetime but no strong opinion to keep that or remove that.

Comment (jsvob) [reply osidb/filters.py:205]: After some thought, the timestamps are unique to the given osidb instance and represent the time the osidb instance learned of the given acknowledgment, not the time when it was actually added into bugzilla. I'll try to reasonably trim the amount of filters in a separate commit. Thanks!

File: openapi.yml
Comment (jsvob) [reply openapi.yml:5940]: Hopefully addressed in commit 3c2d2dd147380fc8924fac4b9d539f85f86f36d4

The swagger page for POST and PUT (when generated **without** commit 4f068286685d90f3de2ea222b5a84348059af6d8 ) has the right `Example Value`s.

POST:

```
{
  "name": "string",
  "affiliation": "string",
  "from_upstream": true,
  "embargoed": true
}
```

and PUT:

```
{
  "name": "string",
  "affiliation": "string",
  "from_upstream": true,
  "embargoed": true,
  "updated_dt": "2023-07-04T18:04:44.202Z"
}
```

`flaw_id` and `id` are still marked as required, but also marked as belonging to `path` as opposed to `query`.

Does that do what you need @JakubFrejlach please? Thanks!

File: osidb/api_views.py
Comment (jsvob) [reply osidb/api_views.py]: Fixed in commit 3c2d2dd147380fc8924fac4b9d539f85f86f36d4

File: osidb/api_views.py
Comment (jsvob) [reply osidb/api_views.py:367]: Fixed in commit 3c2d2dd147380fc8924fac4b9d539f85f86f36d4 except for the new mixin. I'd prefer to do the mixin in a separate MR once we get an approval on this MR because of potential complexity, because it mostly depends on whether we agree on PR265, and because of the iterative approach we discussed two weeks ago.

File: osidb/tests/test_endpoints.py
Comment (jsvob) [reply osidb/tests/test_endpoints.py]: Fixed in commit 3c2d2dd147380fc8924fac4b9d539f85f86f36d4 (hopefully, see discussion thread https://github.com/RedHatProductSecurity/osidb/pull/265#discussion_r1250297359 )

File: osidb/api_views.py
Comment (jsvob) [reply osidb/api_views.py:367]: Fixed in commit 3c2d2dd147380fc8924fac4b9d539f85f86f36d4

File: osidb/serializer.py
Comment (jsvob) [reply osidb/serializer.py:627]: Fixed in commit 3c2d2dd147380fc8924fac4b9d539f85f86f36d4

File: osidb/serializer.py
Comment (jsvob) [reply osidb/serializer.py]: Fixed in commit 3c2d2dd147380fc8924fac4b9d539f85f86f36d4 (hopefully, see discussion thread https://github.com/RedHatProductSecurity/osidb/pull/265#discussion_r1250297359 ). Thanks!

File: osidb/filters.py
Comment (jsvob) [reply osidb/filters.py:205]: Fixed in commit 627341b808e39ae2bd298779e83eabf3aab3ca9a

**Comment by jsvob** [comment]
> This PR should have an entry in CHANGELOG because it contains changes visible to customers.

Fixed in commit 4f068286685d90f3de2ea222b5a84348059af6d8 but perhaps in a way you might not approve of. IDK of a better way though. Should be addressed in OSIDB-1004 IMO. Open to suggestions. Thanks!

File: osidb/filters.py
Comment (osoukup) [reply osidb/filters.py:205]: > After some thought, the timestamps are unique to the given osidb instance and represent the time the osidb instance learned of the given acknowledgment, not the time when it was actually added into bugzilla. I'll try to reasonably trim the amount of filters in a separate commit. Thanks!

Not sure if I understand this correctly but the timestamps should be inherited from the flaw. Bugzilla is still the authoritative source of the data so OSIDB should just mirror it.

Comment (jsvob) [reply osidb/filters.py:205]: Aha, thank you for the correction. `convertors.py` literally contains `created_dt=self.flaw_bug["creation_time"]` when setting `created_dt` of a `FlawAcknowledgment`.

In that light, it's IMO still right to remove `FlawAcknowledgment`'s filtering based on `created_dt`/`updated_dt` because `Flaw` can already be filtered based on `created_dt`/`updated_dt`.

Comment (osoukup) [reply osidb/filters.py:205]: It is not necessary now but in the future it might be useful. I would leave it there. I am not sure if exact match is useful for the timestamps but generally I see the timestamps filtering handy.

Comment (jsvob) [reply osidb/filters.py:205]: In that case, reverted commit 627341b808e39ae2bd298779e83eabf3aab3ca9a in commit d0354a6c6b5652baaee4919d2a446f020b056bc2 and the end result of this conversation is no change.

File: osidb/api_views.py
Comment (jobselko) [inline osidb/api_views.py:397]: Please be more descriptive in docstrings (but still brief). I mean - Which flaw should be gotten? The one associated with a given flaw acknowledgment. If a docstring is too general, it has no value for people who read it.

You can also omit `Similar to FlawView.get_object()` as it is useless.

File: osidb/api_views.py
Comment (jobselko) [inline osidb/api_views.py:407]: Please add a docstring with a description why we need this (e.g. similar to `FlawReferenceView`'s `get_queryset`.

File: osidb/api_views.py
Comment (jobselko) [inline osidb/api_views.py:408]: Remove this condition because we use `drf-spectacular` and not `drf-yasg`.

**Review by jobselko — CHANGES_REQUESTED** [review]

File: osidb/api_views.py
Comment (jobselko) [reply osidb/api_views.py:367]: Add "common mixin" - it was not said that we should do it in this PR. I think it was only a suggestion for the future :).

File: osidb/api_views.py
Comment (Elkasitu) [reply osidb/api_views.py:408]: This is actually a common problem to [both libraries](https://drf-spectacular.readthedocs.io/en/latest/faq.html#my-get-queryset-depends-on-some-attributes-not-available-at-schema-generation-time) and it looks like `drf-spectacular` might've been inspired by `drf-yasg` and I assume the reason for it being needed here is because of the nesting.

File: osidb/api_views.py
Comment (jobselko) [reply osidb/api_views.py:371]: If BBSync is done later, leaving the tests commented out makes sense.

File: osidb/tests/test_endpoints.py
Comment (jobselko) [reply osidb/tests/test_endpoints.py:1641]: If BBSync is done later, leaving the tests commented out makes sense.

File: osidb/api_views.py
Comment (jobselko) [reply osidb/api_views.py:408]: Thanks for the clarification. In that case, we should keep that condition with the link you mentioned to explain why it is important.

Comment (jsvob) [reply osidb/api_views.py:408]: Will fix the comment.

The schema can be generated without this condition, but it generates a warning.

File: osidb/api_views.py
Comment (jsvob) [reply osidb/api_views.py]: Fixed in commits 09866d24b5b72af9f19ea30c518961a04686bd1b and eaaa28d2c2b1ea423c782b4745faa28b97bf3e7d

File: osidb/api_views.py
Comment (jsvob) [reply osidb/api_views.py:413]: Fixed in commits 09866d24b5b72af9f19ea30c518961a04686bd1b and eaaa28d2c2b1ea423c782b4745faa28b97bf3e7d

File: osidb/api_views.py
Comment (jsvob) [reply osidb/api_views.py:417]: Fixed in commits 09866d24b5b72af9f19ea30c518961a04686bd1b and eaaa28d2c2b1ea423c782b4745faa28b97bf3e7d

File: osidb/api_views.py
Comment (jsvob) [reply osidb/api_views.py]: Fixed in commit 09866d24b5b72af9f19ea30c518961a04686bd1b

File: osidb/api_views.py
Comment (jsvob) [reply osidb/api_views.py]: Fixed in commits 09866d24b5b72af9f19ea30c518961a04686bd1b and eaaa28d2c2b1ea423c782b4745faa28b97bf3e7d

File: osidb/api_views.py
Comment (jsvob) [reply osidb/api_views.py:426]: Fixed in commits 09866d24b5b72af9f19ea30c518961a04686bd1b and eaaa28d2c2b1ea423c782b4745faa28b97bf3e7d

File: osidb/serializer.py
Comment (jsvob) [reply osidb/serializer.py]: Fixed in commit 09866d24b5b72af9f19ea30c518961a04686bd1b

File: osidb/tests/test_endpoints.py
Comment (jsvob) [reply osidb/tests/test_endpoints.py:1601]: Fixed in commits c1ecc5d3479860137d8571768172bb620efd4360 and ccc82af3c0f5074bca8229906d82beeb427dcffe

**Comment by jsvob** [comment]
There are many commits now, so that I could reply to feedback in a way that's IMO clearer than amend and force push. I'll squash the commits before merge so that there are 3 commits:

- The main part of the work
- Fix of comments in FlawCommentView
- Fix of request body in test_flawreference_create

File: osidb/api_views.py
Comment (jobselko) [inline osidb/api_views.py:1]: Could you please add a new commit (probably the 5th when you are squashing your final work) to fix [this problem](https://drf-spectacular.readthedocs.io/en/latest/faq.html#my-get-queryset-depends-on-some-attributes-not-available-at-schema-generation-time) with `get_queryset()`  for `FlawReference`?

I checked that spectacular reported this:
`Warning #9: FlawReferenceView: Failed to obtain model through view's queryset due to raised exception. Prevent this either by setting "queryset = Model.objects.none()" on the view, checking for "getattr(self, "swagger_fake_view", False)" in get_queryset() or by simply using @extend_schema. (Exception: 'flaw_id')`

I overlooked it when I was working on FlawReference. Thank you!

**Review by jobselko — CHANGES_REQUESTED** [review]

File: osidb/api_views.py
Comment (jsvob) [reply osidb/api_views.py:1]: Interactively rebased, commits pruned, and your suggestion added as an extra commit. Thanks!

**Comment by jsvob** [comment]
@JakubFrejlach There are backward compatibility errors in CI Meta checks:

> Backward compatibility errors (5):
> error at head/openapi.yml, in API GET /osidb/api/v1/flaws/{flaw_id}/references for the 'path' request parameter 'flaw_id', the type/format was changed from 'string'/'none' to 'string'/'uuid' [request-parameter-type-changed]. 
> ...

This is caused by commit 931021ce7351ddcfc2a0a1c172d5f82a1609a58c. How to resolve it, please?

File: osidb/api_views.py
Comment (jsvob) [reply osidb/api_views.py:1]: @jobselko Reverting the fix you requested. This is blocking other work (OSIDB-1004) and it's out of scope. Please file a separate task if necessary.

File: osidb/api_views.py
Comment (jsvob) [reply osidb/api_views.py:381]: This is blocking other work (OSIDB-1004) and it seems the concerns have been addressed. If a common mixin is still needed, please open a separate Jira, or I'll try to implement it as a part of future similar work (OSIDB-1006 potentially).

File: openapi.yml
Comment (jsvob) [reply openapi.yml:6018]: This is blocking other work (OSIDB-1004) and it seems the concerns have been addressed.

**Comment by jsvob** [comment]
FTR, final changes requested by @jobselko are out of scope. Other feedback has been addressed. Also see these discussion threads for more context why I'm merging this:

- https://github.com/RedHatProductSecurity/osidb/pull/265#discussion_r1265830310
- https://github.com/RedHatProductSecurity/osidb/pull/265#discussion_r1265830183

**Comment by jsvob** [comment]
@jobselko @JakubFrejlach Your reviews are blocking the merge. Please approve or provide actionable feedback.

**Comment by jobselko** [comment]
Unfortunately, Jakub is on PTO now, so we will have to wait until he is back.

Regarding the issue with `get_queryset()` for `FlawReference,` I will look at it in the meantime. It could be an easy fix considering that it works for `FlawAcknowledgment.`

**Comment by jsvob** [comment]
@jobselko I see nothing in team calendar. OK then. I dismissed @JakubFrejlach 's feedback. Please approve from your side. Your request is out of scope. I'm blocked.

**Review by jobselko — APPROVED** [review]
I approve this PR, but I would prefer @Elkasitu to review and approve it too (as we agreed on meeting a few weeks ago).

IMHO, this PR is not so urgent and can wait until @JakubFrejlach is back. If it was a necessary bug fix, then I understand. But in this situation, I find it rude to kick somebody off from the review because he is on PTO.

File: osidb/tests/test_endpoints.py
Comment (Elkasitu) [inline osidb/tests/test_endpoints.py:1598]: I'm not sure what this change accomplishes, perhaps the bit that removes the flaw as required param is in another commit?

**Review by Elkasitu — APPROVED** [review]
Should be squashed, not sure if that is your intention so just a reminder.

In the future, should we do the bbsync part before the API implementation? This would avoid implementing a half-baked API that has only some working methods instead of all? Not sure of the feasibility, just a suggestion

File: osidb/tests/test_endpoints.py
Comment (jsvob) [reply osidb/tests/test_endpoints.py:1598]: This is in a separate commit with the message "Do not require flaw value in body, because it is already provided in URL".

Jakub [requested](https://github.com/RedHatProductSecurity/osidb/pull/265#discussion_r1250405741) that this change be done for FlawAcknowledgments.

Jitka [requested](https://github.com/RedHatProductSecurity/osidb/pull/265#discussion_r1263797140) other FlawReference fixes in this PR to mirror how FlawAcknowledgments are handled.

Jitka [requested](https://github.com/RedHatProductSecurity/osidb/pull/265#discussion_r1262218968) that this change is in a separate commit.

**Comment by jsvob** [comment]
> Should be squashed, not sure if that is your intention so just a reminder.

@Elkasitu My intention was **not** to squash these 4 commits. One of the commits will be reverted in a future PR (enabling write actions on the api). Two other commits are for changes in other code so that the approach is updated to match the new FlawAcknowledgments (comment updates mostly). @jobselko also requested multiple times that such changes should be in separate commits. Should the commits have different messages or is there any other concern?

> In the future, should we do the bbsync part before the API implementation? This would avoid implementing a half-baked API that has only some working methods instead of all? Not sure of the feasibility, just a suggestion

Yes, I agree and will do that next time.

The current approach is for 3 reasons

- I wasn't aware at this point that the data migration should have already been in place together with bbsync at this point. Will do better next time.
- This PR was a time sink for folks for long enough. I thought it would be better to get it merged and out of sight than reordering the other work.
- My original intention was not to have this as a user-facing issue and I wanted to just merge the API, then bbsync. Jitka didn't want the `technical` label and to add a CHANGELOG entry. That made it a user-facing issue in my opinion. And it's not ready for usage. That added the complication of disabling write actions (in a separate commit). If I get hit by a bus and its forgotten, with write actions enabled, we'd start seeing data loss in production bugzilla once prod bbsync was enabled, hence the caution and extra commit.

Our options with _this_ PR are IMO

- Postponing this, implementing bbsync first, and then removing commit 058f6c7ab89725dc9f8aec001c62294da917b258 from this PR. 
- Merging this, then continuing with bbsync, then enabling write api disabled in commit 058f6c7ab89725dc9f8aec001c62294da917b258.

What do you prefer?

File: osidb/tests/test_endpoints.py
Comment (Elkasitu) [reply osidb/tests/test_endpoints.py:1598]: I mean that this commit does not do what it says it does "Do not require flaw value in body, because it is already provided in URL", the commit simply modifies a test, does not change anything about the actual code, given the commit message I would imagine some sort of change to the serializer.

Furthermore, shouldn't passing the flaw id parameter return a 400 response? That might be worthy of a bug ticket in JIRA

Comment (jobselko) [reply osidb/tests/test_endpoints.py:1598]: @Elkasitu The code change happened in https://github.com/RedHatProductSecurity/osidb/pull/256/commits/15c9664494e00bd2a49143c97f14e6f6a70f1e61 in `osidb/serializer.py`. At that time, I forgot to update the test. And maybe, I forgot to update even more tests. I will check it and create another PR with FlawReference fixes, hopefully the last one :)

I agree that the used commit name ("Do not require flaw value in body, because it is already provided in URL".) is a bit confusing.

Comment (jsvob) [reply osidb/tests/test_endpoints.py:1598]: Aha, I meant "Change the test to not provide the flaw value to be inside body, because it is already provided in the URL." I saw the test as something that effectively places requirements on how stuff should work (test should pass). Would the revised commit message in the first sentence work?

Theoretically, yes, it should return a 400, but that would IMO require changes I'm not sure are short enough for us to want. The reason the extra "flaw" in body is not detected (and in fact, always ignored) is that the viewset's `get_serializer()` injects the `flaw` from URL into the body, overwriting any potential value already in the body. The fix to return a 400 would probably be to have an extra condition in `get_serializer()` to detect the extra `flaw` in body and return a 400 response. Shall I file the ticket?

Comment (jsvob) [reply osidb/tests/test_endpoints.py]: Removing the commit removes the bad message.

**Pull request merged**
